src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -334,24 +334,102 @@ private Array ReadArray(DataKind kind)
                     break;
             }
 
-            Type elementType = this.ReadType();
+            var elementKind = (DataKind)_reader.ReadByte();
+
+            // optimization for primitive type array
+            Type elementType;
+            if (s_reverseTypeMap.TryGetValue(elementKind, out elementType))
+            {
+                return this.ReadPrimitiveTypeArrayElements(elementType, elementKind, length);
+            }
+
+            // custom type case
+            elementType = this.ReadType(elementKind);
```

Comment:
Wouldn't it be more efficient to do something like:

```
switch (kind)
{
   case DataKind.Int8: 
      array = new sbyte[length];
      for (int i = 0; i < array.Length; i++) array[i] = ReadSByte();
      break;

   case DataKind.UInt8: 
      array = new byte[length];
      for (int i = 0; i < array.Length; i++) array[i] = ReadByte();
      break;
...
}
```

That is, specialize the entire array creation and initialization by element kind?